### PR TITLE
Add actions for sending subcell reconstruction data

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Exactly what the name says :D

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nilsdeppe** [comment]
Rebased on develop to get the latest action changes. I squashed in the changes since nobody has reviewed this yet

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+#include "Time/Tags.hpp"
+#include "Time/TimeStepId.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace evolution::dg::subcell::Actions {
+/*!
+ * \brief Sets the local data form the relaxed discrete maximum principle
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp]: "form" -> "from" ?

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+                      std::move(data)});
+      }
+    }
+    return {std::move(box)};
+  }
+};
+
+/*!
+ * \brief Receive the subcell data from our neighbor, and accumulate the data
+ * form the relaxed discrete maximum principle troubled-cell indicator.
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp]: "form" -> "from"

File: tests/Unit/Evolution/DgSubcell/Actions/Test_ReconstructionCommunication.cpp [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+  // direction.
+  //
+  // In 1d there aren't any projections to test, and in 3d we only have 1
+  // element in the z-direction.
+  DirectionMap<Dim, Neighbors<Dim>> neighbors{};
+  ElementId<Dim> self_id{};
+  ElementId<Dim> east_id{};
+  ElementId<Dim> south_id{};  // not used in 1d
+  OrientationMap<Dim> orientation{};
+  // Note:
```
Comment (fmahebert) [inline tests/Unit/Evolution/DgSubcell/Actions/Test_ReconstructionCommunication.cpp]: don't leave me hanging Nils

**Comment by nilsdeppe** [comment]
Rebased and pushed a fixup. Thanks for the review :D

**Review by fmahebert — COMMENTED** [review]
👍 you can squash (I apologize, should have asked for these to be directly squashed)

**Comment by nilsdeppe** [comment]
:+1: squashed :)

**Review by fmahebert — DISMISSED** [review]

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+ *   variables to the neighbors. This will end up depending on the
+ *   reconstruction scheme used, so it'll take a bit of infrastructure
+ *   development.
+ * - In the future we will need to send the cell-centered fluxes to do
+ *   high-order FD without additional reconstruction being necessary.
+ *
+ * GlobalCache:
+ * - Uses:
+ *   - `ParallelComponent` proxy
+ * DataBox:
```
Comment (kidder) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp:80]: need a blank line for doc to render correctly

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+    const Mesh<Dim>& dg_mesh = db::get<::domain::Tags::Mesh<Dim>>(box);
+    const Mesh<Dim>& subcell_mesh = db::get<Tags::Mesh<Dim>>(box);
+    const Element<Dim>& element = db::get<::domain::Tags::Element<Dim>>(box);
+    const size_t ghost_zone_size =
+        Metavariables::SubcellOptions::ghost_zone_size(box);
+    // Note: we currently always send the conserved variables, but will need to
+    // generalize that to allow the user to choose what to send. For example,
+    // with relativistic hydrodynamics it's desirable to reconstruct
+    //    u_i = u^0(lapse * v_i - shift_i)
```
Comment (kidder) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp]: Your equation is incorrect as:
- u^i = u^0 (lapse * v^i - shift^i) 
- u^0 = W/lapse
- u_i = W * v_i   
- u_0 = W (-lapse + v_i * shift^i)

I think it is correct that you want to reconstruct u_i as W^2 = 1 + gamma^ij u_i u_j = 1 + W^2 v^2

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+      const auto& orientation = neighbors_in_direction.orientation();
+      const auto direction_from_neighbor = orientation(direction.opposite());
+      // If we are periodic in this direction don't send any data.
+      ASSERT(neighbors_in_direction.size() == 1,
+             "AMR is not yet supported when using DG-subcell. Note that this "
+             "condition could be relaxed to support AMR only where the "
+             "evolution is using DG without any changes to subcell.");
+
+      for (const ElementId<Dim>& neighbor : neighbors_in_direction) {
+        std::vector<double> scl_data_to_send{};
```
Comment (kidder) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp]: scl -> subcell

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    }
+    // Optimization note: could save a copy+allocation if we moved
+    // all_sliced_data when possible before sending.
+    const DirectionMap<Dim, std::vector<double>> all_sliced_data = slice_data(
+        db::mutate_apply(GhostDataMutator{}, make_not_null(&box)),
+        subcell_mesh.extents(), ghost_zone_size, directions_to_slice);
+
+    auto& receiver_proxy =
+        Parallel::get_parallel_component<ParallelComponent>(cache);
+    const NeighborData& local_neighbor_data =
```
Comment (kidder) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp:155]: hasn't this been called self_neighbor_data in other places?

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+    }
+
+    // Now that we have received all the data, copy it over as needed.
+    FixedHashMap<maximum_number_of_neighbors(Dim), Key,
+                 std::tuple<Mesh<Dim - 1>, std::optional<std::vector<double>>,
+                            std::optional<std::vector<double>>, ::TimeStepId>,
+                 boost::hash<Key>>
+        received_data = std::move(inbox[current_time_step_id]);
+    inbox.erase(current_time_step_id);
+    static_assert(not Metavariables::local_time_stepping);
```
Comment (kidder) [inline src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp]: you already asserted this on line 263

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash immediately

File: src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp (hunk 7 above)
Comment (nilsdeppe) [reply src/Evolution/DgSubcell/Actions/ReconstructionCommunication.hpp:155]: Maybe, I don't remember off the top of my head.

**Comment by nilsdeppe** [comment]
I've rebased and squashed. Thanks for the reviews :)

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**